### ISSUE TITLE
VertexBatchTest.Locking failing locally

### ISSUE THREAD

**Issue by Barsonax** [AdamsLair/duality#719]
### Summary
The VertexBatchTest.Locking is failing for me locally

```
Expected: False
  But was:  True

   at Duality.Tests.Drawing.VertexBatchTest.Locking() in C:\Git\duality\Test\Core\Drawing\VertexBatchTest.cs:line 90
```

### How to reproduce
- Run the test
- It fails (well or not because it does not fail on the CI)

### Workaround
- Ignore it as it only happens locally

### Analysis
- Likely happens due to something environment specific. The question is what?

**Comment by ilexp** [comment]
The specific part of the test that is failing is [this](https://github.com/AdamsLair/duality/blob/67823f4071ba2e7cca3063087bd5d8bd6c244911/Test/Core/Drawing/VertexBatchTest.cs#L84-L90):

```csharp
// Make sure that our locks released properly, i.e. allowing the array to be garbage collected
WeakReference weakRefToLockedData = new WeakReference(typedBatch.Vertices.Data);
Assert.IsTrue(weakRefToLockedData.IsAlive);
typedBatch = null;
abstractBatch = null;
GC.Collect();
Assert.IsFalse(weakRefToLockedData.IsAlive);
```

Which is dependent on garbage collection behavior, or more specifically the assumption that `GC.Collect` will free the vertex data array synchronously before returning. Could this be related? Are there ways to ensure this across runtime / GC versions?

**Comment by Barsonax** [comment]
It does not do it synchronously:
https://stackoverflow.com/questions/748777/run-gc-collect-synchronously

Also I think depending on the GC like this is very error prone as it can change between runtime or config settings.

**Comment by ilexp** [comment]
> Also I think depending on the GC like this is very error prone as it can change between runtime or config settings.

It is, but since this part of the test is essentially a GC test, there's not really a way around that.

> It does not do it synchronously:
> https://stackoverflow.com/questions/748777/run-gc-collect-synchronously

That's likely the issue then. Could be worth a try to use [this](https://docs.microsoft.com/en-us/dotnet/api/system.gc.collect?view=netframework-4.8#System_GC_Collect_System_Int32_System_GCCollectionMode_System_Boolean_) overload instead and also wait for pending finalizers afterwards.

**Comment by ilexp** [comment]
Adjusted the test as described above. @Barsonax Can you check whether the issue is fixed on your machine?

**Comment by Barsonax** [comment]
Will check when iam at my pc

**Comment by Barsonax** [comment]
It did not fix the issue

**Comment by Barsonax** [comment]
Only happens in Debug mode, under release mode the test runs succesfully.

So this error can be fixed by skipping this test in debug mode (no sense in running it if it always fails). In nunit this can be done by making a custom `NUnitAttribute` and implementing `IApplyToTest`. Ofcourse a reasonable message should be given explaining that it does not run in debug.

**Comment by Barsonax** [comment]
Fixed in #719
